Lib/os.py

```diff
@@ -485,58 +489,89 @@ def fwalk(top=".", topdown=True, onerror=None, *, follow_symlinks=False, dir_fd=
     def _fwalk(topfd, toppath, isbytes, topdown, onerror, follow_symlinks):
         # Note: This uses O(depth of the directory tree) file descriptors: if
         # necessary, it can be adapted to only require O(1) FDs, see issue
-        # #13734.
+        # bpo-13734 (gh-57943).
+        stack = [(_WalkAction.WALK, (topfd, toppath))]
+        fd_stack = []
```

Comment:
Could you store the fd to close as the second argument of stack tuples, like `(_WalkAction.CLOSE, fd)`, rather than using a separate stack? It would mean you'd need to traverse `stack` in your `finally:` block, but that's probably OK?

Reply:
We could with a couple of caveats. I actually originally did that, but then realized that the fd wouldn't get closed if an exception is raised immediately after it's popped from the stack (i.e. it won't get handled in the `finally`). It also means iterating through a longer list and checking the action type for `_WalkAction.CLOSE` in the `finally` - like you said probably not a big deal, but having `fd_stack` does make this cleanup step simpler.

See the commit where I changed this for the difference: https://github.com/python/cpython/pull/100347/commits/f0f9330f4c2a6bbb24957002f1442ca135022c37
Note, even before this commit, to make sure something like a `KeyboardException` doesn't leave the fd open, we probably need change the code to first check the top of the stack for an fd, close it, and then pop the item:
```python
action, value = stack[-1]
...
stack.pop()
```
Maybe this isn't critical, but I figured a second list was simple enough. Lmk if you still think it's worth getting rid of `fd_stack`.

Reply:
I think I was maybe looking at this wrong. By popping after calling `close`, we risk calling `close` twice on the same fd. Anyway, switched to using one `stack` for now. That simplifies a little and optimizes for the case where we don't run into an exception. Iterating over the entire stack in the `except` should be an acceptable cost I think